### PR TITLE
ROC-2477 Add template

### PR DESCRIPTION
**Pull request by timja-kainos**
https://tools.hmcts.net/jira/browse/ROC-2477

### CODE REVIEW

File: src/main/resources/staff/templates/document/settlementAgreement.html [hunk 1]
```diff
@@ -1 +1,190 @@
...
+  <div class="section-body">
+    <table>
+      <tr>
+        <td>Agreement:</td>
+        <td>
+          {{ acceptedOffer }}
+        </td>
+      </tr>
+      <tr>
+        <td>Payment date:</td>
```
Comment (rlewan) [inline src/main/resources/staff/templates/document/settlementAgreement.html]: Don't we need to output the date here as well?

File: cmc-domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/offers/Settlement.java [hunk 2]
```diff
@@ -33,11 +34,28 @@ public void reject(MadeBy party) {
...
+    @JsonIgnore
+    public PartyStatement getLastStatementWithOffer() {
+        if (partyStatements.isEmpty()) {
+            throw new IllegalSettlementStatementException(NO_STATEMENTS_MADE);
+        }
+
+        List<PartyStatement> tmpList = new ArrayList<>(partyStatements);
+        Collections.reverse(tmpList);
+
+        return tmpList.stream()
```
Comment (timja-kainos) [inline cmc-domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/offers/Settlement.java]: will add a test for this

File: src/main/resources/staff/templates/document/settlementAgreement.html [hunk 3]
```diff
@@ -1 +1,190 @@
...
+      <tr>
+        <td>Full Name:</td>
+        <td>
+          {{ defendant.fullName }}
+        </td>
+      </tr>
+      <tr>
+        <td>Address:</td>
+        <td>
+          {{ defendant.address.line1 }}
```
Comment (timja-kainos) [inline src/main/resources/staff/templates/document/settlementAgreement.html]: defendant address is currently the claimant address for some reason, tracking it down

File: cmc-domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/offers/Settlement.java [hunk 4]
```diff
@@ -33,11 +34,27 @@ public void reject(MadeBy party) {
...
     PartyStatement getLastStatement() {
         if (partyStatements.isEmpty()) {
-            throw new IllegalSettlementStatementException("No statements have yet been made during that settlement");
+            throw new IllegalSettlementStatementException(NO_STATEMENTS_MADE);
         }
         return partyStatements.get(partyStatements.size() - 1);
     }
 
+    @JsonIgnore
+    public PartyStatement getLastStatementWithOffer() {
```
Comment (timja-kainos) [inline cmc-domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/offers/Settlement.java]: not sure whether to do this or just do:
```
return partyStatements.get(partyStatements.size() - 2);
```

Comment (rlewan) [reply cmc-domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/offers/Settlement.java]: I think this is fine. I'd rename the method to just `getLastOffer` though.

File: cmc-domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/offers/Settlement.java [hunk 5]
```diff
@@ -33,11 +34,27 @@ public void reject(MadeBy party) {
...
+        if (partyStatements.isEmpty()) {
+            throw new IllegalSettlementStatementException(NO_STATEMENTS_MADE);
+        }
+
+        List<PartyStatement> tmpList = new ArrayList<>(partyStatements);
+        Collections.reverse(tmpList);
+
+        return tmpList.stream()
+            .filter((partyStatement -> partyStatement.getOffer().isPresent()))
```
Comment (rlewan) [inline cmc-domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/offers/Settlement.java]: I know this is correct but I think I'd still filter on statement type, not offer being present. It's more strict, may be easier to detect some bugs in the future if any show up.

File: cmc-domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/offers/Settlement.java (hunk 4 above)
Comment (rlewan) [reply cmc-domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/offers/Settlement.java]: Or `getLastOfferStatement`.

Comment (timja-kainos) [reply cmc-domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/offers/Settlement.java]: done the second one

File: cmc-domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/offers/Settlement.java (hunk 5 above)
Comment (timja-kainos) [reply cmc-domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/offers/Settlement.java]: done

File: cmc-domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/offers/Settlement.java [hunk 6]
```diff
@@ -33,11 +34,27 @@ public void reject(MadeBy party) {
...
+        if (partyStatements.isEmpty()) {
+            throw new IllegalSettlementStatementException(NO_STATEMENTS_MADE);
+        }
+
+        List<PartyStatement> tmpList = new ArrayList<>(partyStatements);
+        Collections.reverse(tmpList);
+
+        return tmpList.stream()
+            .filter((partyStatement -> partyStatement.getType() == StatementType.OFFER))
+            .findFirst()
```
Comment (geek-soft) [inline cmc-domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/offers/Settlement.java]: Maybe we could sort it here by using .sorted(Collections.reverseOrder()) and then take first element. That could help us avoid using nasty Collections.reverse().

File: cmc-domain-model/src/test/java/uk/gov/hmcts/cmc/domain/models/offers/SettlementTest.java [hunk 7]
```diff
@@ -123,4 +124,30 @@ public void partyIsNotAllowedToRejectOfferWhenTheOnlyOfferWasMadeByThemselves()
...
+    @Test
+    public void getLastOfferStatementShouldGiveLastStatementThatIsAnOffer() {
+        settlement.makeOffer(offer, MadeBy.CLAIMANT);
+        settlement.reject(MadeBy.DEFENDANT);
+
+        Offer counterOffer = SampleOffer.builder()
+            .withContent(COUNTER_OFFER)
+            .build();
+
+        settlement.makeOffer(counterOffer, MadeBy.DEFENDANT);
```
Comment (geek-soft) [inline cmc-domain-model/src/test/java/uk/gov/hmcts/cmc/domain/models/offers/SettlementTest.java]: Is defendant making counter offer? Shouldn't that be claimant?

File: src/main/java/uk/gov/hmcts/cmc/claimstore/controllers/SupportController.java [hunk 8]
```diff
@@ -64,7 +70,7 @@ public void resendStaffNotifications(
     ) throws ServletRequestBindingException {
 
         Claim claim = claimRepository.getByClaimReferenceNumber(referenceNumber)
-            .orElseThrow(() -> new NotFoundException("Claim " + referenceNumber + " does not exist"));
+            .orElseThrow(() -> new NotFoundException(CLAIM + referenceNumber + " does not exist"));
```
[the remark this answers is not in the document]
Comment (timja-kainos) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/controllers/SupportController.java]: Sonar asked for it... I didn’t choose too

File: cmc-domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/offers/Settlement.java (hunk 6 above)
Comment (timja-kainos) [reply cmc-domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/offers/Settlement.java]: That only works if there’s a comparator we rely on natural ordering. I tried this first

Comment (timja-kainos) [reply cmc-domain-model/src/main/java/uk/gov/hmcts/cmc/domain/models/offers/Settlement.java]: That only works if there’s a comparator we rely on natural ordering. I tried this first

File: cmc-domain-model/src/test/java/uk/gov/hmcts/cmc/domain/models/offers/SettlementTest.java (hunk 7 above)
Comment (timja-kainos) [reply cmc-domain-model/src/test/java/uk/gov/hmcts/cmc/domain/models/offers/SettlementTest.java]: Based on the frontend yes, but the backend allows both way.

**Pull request merged**